testing/async/mixin.h

```diff
@@ -0,0 +1,645 @@
+//    // the prefix and postfix tuples. These tuples contains instances of any
+//    // additional arguments to pass to the algorithm. The tuple/index_sequence
+//    // pattern is used to support the "no extra args" overload, since the
+//    // parameter pack expansion will do exactly what we want in all cases.
+//    // Prefix args are included here (but not for invoke_reference) to allow
+//    the
+//    // test framework to change the execution policy.
+//    // This method must return an event or future.
+//    template <typename PrefixArgTuple,
+//              std::size_t... PrefixArgIndices,
+//              typename PostfixArgTuple,
+//              std::size_t... PostfixArgIndices>
+//    static auto invoke_async(PrefixArgTuple &&prefix_tuple,
+//                             std::index_sequence<PrefixArgIndices...>,
+//                             input_type const &input,
+//                             output_type &output,
+//                             PostfixArgTuple &&postfix_tuple,
+//                             std::index_sequence<PostfixArgIndices...>)
+//    {
+//      output.resize(input.size());
+//      auto e = thrust::async::exclusive_scan(
+//        std::get<PrefixArgIndices>(THRUST_FWD(prefix_tuple))...,
+//        input.cbegin(),
+//        input.cend(),
+//        output.begin(),
+//        std::get<PostfixArgIndices>(THRUST_FWD(postfix_tuple))...);
+//      return e;
+//    }
+//
+//    //-------------------------------------------------------------------------
+//    // testing::async::mixin::compare_outputs::assert_equal_quiet
+//    //
+//    // Wait on and validate the event/future (usually with TEST_EVENT_WAIT /
+//    // TEST_FUTURE_VALUE_RETRIEVAL), then check that the reference output
+//    // matches the testing output.
+//    template <typename EventType>
+//    static void compare_outputs(EventType &e,
+//                                output_type const &ref,
+//                                output_type const &test)
+//    {
+//      TEST_EVENT_WAIT(e);
+//      ASSERT_EQUAL_QUIET(ref, test);
+//    }
+// };
+// ```
+//
+// Similar invokers with slight tweaks are used in other
+// async/exclusive_scan/*.cu tests.
+
+namespace testing
+{
+namespace async
+{
+namespace mixin
+{
+
+//------------------------------------------------------------------------------
+namespace input
+{
+
+// TODO it'd be nice to specify a lambda expression that'd replace the call to
+// thrust::sequence when desired.
+template <typename value_type>
+struct device_vector
+{
+  using input_type = thrust::device_vector<value_type>;
+
+  static input_type generate_input(std::size_t num_values)
+  {
+    input_type input(num_values);
+    thrust::sequence(input.begin(),
+                     input.end(),
+                     static_cast<value_type>(1.5),
+                     static_cast<value_type>(1));
+    return input;
+  }
+};
+
+template <typename value_type>
+struct counting_iterator_0
```

Comment:
Could this be `_starts_at_0` or `_from_0`, etc?

Reply:
Changed to `_from_0`.